Add breaking change to guild scheduled events and more details for guild scheduled events feature (#4113)
hammerandchisel/discord-api-docs: docs/resources/Guild_Scheduled_Event.md

| entity_id             | ?snowflake                                                                                                                     | the id of an entity associated with a guild scheduled event                                         |
| entity_metadata **    | ?[entity metadata](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-entity-metadata)   | additional metadata for the guild scheduled event                                                   |
| creator?              | [user](#DOCS_RESOURCES_USER/user-object) object                                                                                | the user that created the scheduled event                                                           |
| user_count?           | integer                                                                                                                        | the number of users subscribed to the scheduled event                                               |

> \* `creator_id` will be null and `creator` will not be included for events created before October 25th, 2021, when the concept of `creator_id` was introduced and tracked.

Reviewer: ```suggestion
> note
> `creator_id` will be null and `creator` will not be included for events created before October 25th, 2021, when the concept of `creator_id` was introduced and tracked.
```

Making this a note will be better for consistency of the docs, and illustrates importance for data model purposes. (Plus I think it looks better)

\** See [field requirements by entity type](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-field-requirements-by-entity-type) to understand the relationship between `entity_type` and the following fields: `channel_id`, `entity_metadata`, and `scheduled_end_time` 

###### Guild Scheduled Event Privacy Level

| Level      | Value | Description                                              |
| ---------- | ----- | -------------------------------------------------------- |
| GUILD_ONLY | 2     | the scheduled event is only accessible to guild members  |

###### Guild Scheduled Event Entity Types

| Type           | Value |
| -------------- | ----- |
| STAGE_INSTANCE | 1     |
| VOICE          | 2     |
| EXTERNAL       | 3     |

###### Field Requirements By Entity Type

> The following matrix describes field requirements based on current entity type.

Reviewer: ```suggestion
The following table shows field requirements based on current entity type.
```

> `non null` : This field is required to be a non-null value

Reviewer 2: ```suggestion
`non-null` : This field is required to be a non-null value
```

Author: Good callout. I'm adding to change log as well, my intention was to draw attention to this temporarily since these docs were only recently added, and then remove the warn blocks at later date so the docs aren't cluttered

> `non-null` : This field is required to be a non-null value

Reviewer: ```suggestion
`value` : This field is required to be a non-null value
```

"Non-null" can appear a little too enticing to a reader at first. Simplifying the wording to simply "value" is way easier to comprehend, as a majority or if not any developer will ideally know what is a null and what is not.

`null`     : This field is required to be null

`-`        : No strict requirements

| Entity Type    | channel_id | entity_metadata | scheduled_end_time |
| -------------- | ---------- | --------------- | ------------------ |
| STAGE_INSTANCE |  non-null  |      null       |         -          |
| VOICE          |  non-null  |      null       |         -          |
| EXTERNAL       |    null    |    non-null *   |      non-null      |

> 

Reviewer: ```
| Type           | channel_id | entity_metadata | scheduled_end_time |
| -------------- | ---------- | --------------- | ------------------ |
| STAGE_INSTANCE |   value    |      null       |         -          |
| VOICE          |   value    |      null       |         -          |
| EXTERNAL       |   null     |      value *    |       value        |
```

Appropriating the difference here with "value" now instead of "non-null." For ease of reference, we've also already established that we're dealing with entity types, so just say type instead of reiterating.

> \* `entity_metadata` with a non-null `location` must be provided

Reviewer: Removed suggestion.

Author: this might be the one place where it's still good to keep the strict `non-null` terminology especially since it's outside of the table

Reviewer 3: I think the same. We should keep that here

Reviewer: Updated original suggestion point

###### Guild Scheduled Event Status

| Type        | Value |
| ----------- | ----- |
| SCHEDULED   | 1     |
| ACTIVE      | 2     |
| COMPLETED * | 3     |
| CANCELED  * | 4     |

\* Once `status` is set to `COMPLETED` or `CANCELED`, the `status` can no longer be updated

###### Valid Guild Scheduled Event Status Transitions

SCHEDULED --> ACTIVE

ACTIVE --------> COMPLETED

SCHEDULED --> CANCELED

###### Guild Scheduled Event Entity Metadata

| Field        | Type                | Description                              |
| ------------ | ------------------- | ---------------------------------------- |

> | location? *  | string              | location of the event (1-100 characters) |

Author: NOTE: the deprecated `speaker_ids` field was included here in the initial publication and was removed in a previous change

| description?         | string                                                                                                                      | the description of the scheduled event                                                     |
| entity_type          | [entity type](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-entity-types)        | the entity type of the scheduled event                                                     |

\* [null](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-entity-metadata) for events with `'entity_type': EXTERNAL` 

## Get Guild Scheduled Event % GET /guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/scheduled-events/{guild_scheduled_event.id#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object}

Get a guild scheduled event. Returns a [guild scheduled event](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object) object on success.

###### Query String Params

> warn

> > The with_user_count query param was introduced on Thursday Nov 18, 2021 after the initial publication of this documentation to remain consitent with other guild scheduled events endpoints

Reviewer: ```suggestion
> The `with_user_count` query param was introduced on Thursday Nov 18, 2021 after the initial publication of this documentation to remain consitent with other guild scheduled events endpoints
```

Referring to fields/properties should respect markdown syntax

###### JSON Params

| Field                | Type                                                                                                                        | Description                                                                                |
| -------------------- | --------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------ |

> | channel_id           | ?snowflake *                                                                                                                | the channel id of the scheduled event.                                                     |

Reviewer 4: This is a breaking change. It should be noted that this is going from a "not required" to a "required null" option for External events. This isn't a huge issue but imo if possible it is cleaner to support `undefined` on the CREATE endpoint.

`| channel_id?           | ?snowflake *    the channel id of the scheduled event. ` Something like this so it supports both undefined and null

| status                | [event status](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-status)             | the status of the scheduled event                                                          |

\* If updating `entity_type` to `EXTERNAL`, `channel_id` [must be set to null](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-field-requirements-by-entity-type)

## Delete Guild Scheduled Event % DELETE /guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/scheduled-events/{guild_scheduled_event.id#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object}

Delete a guild scheduled event. Returns a `204` on success.

## Get Guild Scheduled Event Users % GET /guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/scheduled-events/{guild_scheduled_event.id#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object}/users

> warn
> A breaking change was introduced for this endpoint on Thursday Nov 18, 2021 after the initial publication of this documentation in which the return type was changed in response to developer feedback. We apologize for the inconvenience and additional work this creates for developers.

> Get a list of users subscribed to a guild scheduled event. Returns a list of [guild scheduled event user](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-user-object) objects on success. Guild member data, if it exists, is included if the `with_member` query parameter is set.

Reviewer 5: With some testing yesterday, I found out that this endpoint is paginated and respected the `after={user_id}` query parameter, will that still be the case? If so, that should be documented.

Author: yes, both `before` and `after` have been part of the api implementation so should be functional already but we've been holding off on officially adding this to docs until post launch since we don't currently use these internally yet. But at this point I think some libraries are starting to use it already anyway so could just do a bit more testing and prob include this as part of this docs update

###### JSON Params

| Field                 | Type                                                                                                                        | Description                                                                                |
| --------------------- | --------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------ |

> | channel_id *          | snowflake                                                                                                                   | the channel id of the scheduled event, set to `null` if changing entity type to `EXTERNAL` |

Reviewer 6: ```suggestion
| channel_id *          | ?snowflake                                                                                                                   | the channel id of the scheduled event, set to `null` if changing entity type to `EXTERNAL` |
```

Clarify the value is nullable, not just optional.

Author: good catch

| scheduled_end_time    | ISO8601 timestamp                                                                                                           | the time when the scheduled event is scheduled to end                                      |
| description           | string                                                                                                                      | the description of the scheduled event                                                     |
| entity_type *         | [event entity type](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-entity-types)  | the entity type of the scheduled event                                                     |
| status                | [event status](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-guild-scheduled-event-status)             | the status of the scheduled event                                                          |

> \* If updating `entity_type` to `EXTERNAL`, `channel_id` [must be set to null](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-field-requirements-by-entity-type)

Reviewer 4: ```suggestion
\* There are some required properties if updating `entity_type` to `EXTERNAL`. `channel_id` [must be set to null](#DOCS_RESOURCES_GUILD_SCHEDULED_EVENT/guild-scheduled-event-object-field-requirements-by-entity-type). A `location` must be provided in the entity_metadata. A `scheduled_end_time` is required to be set.
```

Would be nice to mention here that when changing to External `location` and `end` time is required alongside `channel_id` must be null. This is because right above this table, docs mention that all properties here are optional. P.S. My wording may not be the best so feel free to change to any wording that you prefer.

Another interesting note not sure if worth documenting though is that when you change from External to voice, `channel_id` is required but `location` is not required to be null, this will be cleared automatically by the api.
